dpnp/dpnp_iface_histograms.py

```diff
@@ -393,44 +454,69 @@ def histogram(a, bins=10, range=None, density=None, weights=None):
 
     a, weights, usm_type = _ravel_check_a_and_weights(a, weights)
 
-    bin_edges, uniform_bins = _get_bin_edges(a, bins, range, usm_type)
+    bin_edges, _ = _get_bin_edges(a, bins, range, usm_type)
 
     # Histogram is an integer or a float array depending on the weights.
     if weights is None:
         ntype = dpnp.dtype(dpnp.intp)
     else:
         ntype = weights.dtype
 
-    # The fast path uses bincount, but that only works for certain types
-    # of weight
-    # simple_weights = (
-    #     weights is None or
-    #     dpnp.can_cast(weights.dtype, dpnp.double) or
-    #     dpnp.can_cast(weights.dtype, complex)
-    # )
-    # TODO: implement a fast path
-    simple_weights = False
-
-    if uniform_bins is not None and simple_weights:
-        # TODO: implement fast algorithm for equal bins
-        pass
+    queue = a.sycl_queue
+    has_fp64 = queue.sycl_device.has_aspect_fp64
+
+    a_bin_dtype, hist_dtype = _align_dtypes(
+        a.dtype, bin_edges.dtype, ntype, has_fp64
+    )
+
+    if a_bin_dtype is None or hist_dtype is None:
+        raise ValueError(
+            f"function '{histogram}' does not support input types "
+            f"({a.dtype}, {bin_edges.dtype}, {ntype}), "
+            "and the inputs could not be coerced to any "
+            "supported types"
+        )
+
+    a_casted = a.astype(a_bin_dtype, order="C", copy=False)
```

Comment:
Because `a` may be `usm_ndarray` (according to docstring), it is not safe to use `a.astype`, since `usm_ndarray` does not implement such a method. Instead `dpctl.tensor.astype` should be used, or support for such input types should not be advertised.

Reply:
Done, replaced with `dpnp.astype(a, ...)`